src/storage/src/hummock/sstable/multi_builder.rs

```diff
@@ -32,33 +44,136 @@ pub struct SealedSstableBuilder {
     pub bloom_filter_size: usize,
 }
 
+/// Consumer for the output of `SstableBuilder::finish`.
+#[async_trait::async_trait]
+pub trait SstableBuilderSealer<W>
+where
+    W: SstableWriter,
+{
+    async fn seal(
+        &self,
+        writer_output: W::Output,
+        sst_meta: SstableMeta,
+        sst_info: SstableInfo,
+        tracker: Option<MemoryTracker>,
+    ) -> HummockResult<SealedSstableBuilder>;
+}
+
+/// Upload the in-memory SST data and metadata all at once.
+pub struct BatchUploadSealer {
+    sstable_store: SstableStoreRef,
+    policy: CachePolicy,
+}
+
+impl BatchUploadSealer {
+    pub fn new(sstable_store: SstableStoreRef, policy: CachePolicy) -> BatchUploadSealer {
+        Self {
+            sstable_store,
+            policy,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SstableBuilderSealer<InMemSstableWriter> for BatchUploadSealer {
+    async fn seal(
+        &self,
+        writer_output: <InMemSstableWriter as SstableWriter>::Output,
+        meta: SstableMeta,
+        sst_info: SstableInfo,
+        tracker: Option<MemoryTracker>,
+    ) -> HummockResult<SealedSstableBuilder> {
+        let bloom_filter_size = meta.bloom_filter.len();
+        let policy = self.policy;
+        let sstable_store = self.sstable_store.clone();
+        let upload_join_handle = tokio::spawn(async move {
```

Comment:
If the only difference is `put_sst` and `finish_put_sst_stream`, we do not need to a trait.
A flag `is_stream` is enough

Reply:
There are a few other differences.

|  batch upload   |  streaming upload  |
|  ----  | ----  |
| data is appended to a buffer | data is written to `SstableStreamingUploader` provided by `SstableStore` |
| `put_sst` is called when all data are written  | `put_sst_stream` should be called when `SstableBuilder` is `new`ed,  `finish_put_sst_stream` is called when all data are written |
| memory usage is proportional to capacity | memory usage is dependent on object store implementation |

Furthermore, we think it'd be better to decouple uploading with SST building. This refactor can achieve it.